src/aosm/azext_aosm/common/registry.py

```diff
@@ -0,0 +1,497 @@
+        Create a list of registry objects from the registries provided by the user.
+
+        :return: A list of registry objects
+        """
+
+        registries: Dict[str, ContainerRegistry] = {}
+        registry_list: List[ContainerRegistry] = []
+
+        logger.debug("Creating registry list from the provided registries")
+
+        for registry in self.image_sources:
+            registry = clean_registry_name(registry)
+
+            # If registry matches the pattern of myacr1.azurecr.io/**,
+            # then create a an instance of the ACR registry class
+            parts = registry.split("/", 1)
+            registry_name = parts[0]
+            registry_namespace = parts[1] if len(parts) > 1 else None
+
+            acr_match = re.match(ACR_REGISTRY_NAME_PATTERN, registry_name)
+
+            if registry_name not in registries:
+                if acr_match:
+                    registries[registry_name] = AzureContainerRegistry(registry_name)
+                else:
+                    registries[registry_name] = UniversalRegistry(registry_name)
+
+                registry_list.append(registries[registry_name])
+
+            if registry_namespace:
+                # Make sure that the namespace ends with a slash
+                if not registry_namespace.endswith("/"):
+                    registry_namespace += "/"
+                registries[registry_name].add_namespace(registry_namespace)
+            else:
+                registries[registry_name].add_namespace("")
+
+        return registry_list
+
+    def get_registry_list(self) -> List[ContainerRegistry]:
+        """
+        Get the list of registries.
+
+        :return: The list of registries
+        """
+        return self.registry_list
+
+    def _get_registries_for_images(
+        self,
+    ) -> Dict[Tuple[str, str], Tuple["AzureContainerRegistry", str]]:
+        """
+        Cycle through all available registries to find images in each.
+
+        :return: A dictionary where keys are (image, version) and values are (registry, namespace)
+        """
+        registry_for_image = {}
+
+        for registry in self.registry_list:
+
+            # We can only do this for ACRs because docker does not have this functionality.
+            if isinstance(registry, AzureContainerRegistry):
+                registry_for_image.update(registry.get_images_in_registry())
+
+        logger.debug("Images found in the ACR registires: %s", registry_for_image)
+        return registry_for_image
+
+    def find_registry_for_image(
+        self, image, version
+    ) -> Union[Tuple["ContainerRegistry", str], Tuple[None, None]]:
+        """
+        Find the registry and namespace for a given image and version.
+
+        :param image: The image to find the registry for
+            This should be stripped of all namespace (e.g nginx, not samples/nginx)
+        :param version: The version of the image
+        :return: The registry and namespace for the image
+        """
+
```

Comment:
We talked about having a fastpath where only when image source was provided. Instead of checking the source (which is a remote call, and therefore slow) we just assume all images are in that single source (as we used to do before supporting multiple sources). This will keep the build command quick in these cases (which we suspect is the majority).

If the image is missing, it will fail in the publish step, and that's OK.